hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/QuotaRepairTask.java

```diff
@@ -0,0 +1,326 @@
+      = new ConcurrentHashMap<>();
+  private final Map<String, CountPair> directoryCountMap
+      = new ConcurrentHashMap<>();
+
+  public QuotaRepairTask(OMMetadataManager metadataManager) {
+    this.metadataManager = metadataManager;
+  }
+  
+  public void repair() throws Exception {
+    LOG.info("Starting quota repair task");
+    prepareAllVolumeBucketInfo();
+
+    IOzoneManagerLock lock = metadataManager.getLock();
+    // thread pool with 3 Table type * (1 task each + 3 thread each)
+    executor = Executors.newFixedThreadPool(12);
+    try {
+      nameBucketInfoMap.values().stream().forEach(e -> lock.acquireReadLock(
+          BUCKET_LOCK, e.getVolumeName(), e.getBucketName()));
+      repairCount();
+    } finally {
+      nameBucketInfoMap.values().stream().forEach(e -> lock.releaseReadLock(
+          BUCKET_LOCK, e.getVolumeName(), e.getBucketName()));
+      executor.shutdown();
+      LOG.info("Completed quota repair task");
+    }
+  }
+  
+  private void prepareAllVolumeBucketInfo() throws IOException {
+    try (TableIterator<String, ? extends Table.KeyValue<String, OmVolumeArgs>>
+        iterator = metadataManager.getVolumeTable().iterator()) {
+
+      OmVolumeArgs omVolumeArgs;
+      while (iterator.hasNext()) {
+        Table.KeyValue<String, OmVolumeArgs> entry =
+            iterator.next();
+        omVolumeArgs = entry.getValue();
+        getAllBuckets(omVolumeArgs.getVolume(), omVolumeArgs.getObjectID());
+      }
+    }
+  }
+  
+  private void getAllBuckets(String volumeName, long volumeId)
+      throws IOException {
+    List<OmBucketInfo> bucketList = metadataManager.listBuckets(
+        volumeName, null, null, Integer.MAX_VALUE, false);
+    for (OmBucketInfo bucketInfo : bucketList) {
+      bucketInfo.incrUsedNamespace(-bucketInfo.getUsedNamespace());
+      bucketInfo.incrUsedBytes(-bucketInfo.getUsedBytes());
+      nameBucketInfoMap.put(buildNamePath(volumeName,
+          bucketInfo.getBucketName()), bucketInfo);
+      idBucketInfoMap.put(buildIdPath(volumeId, bucketInfo.getObjectID()),
+          bucketInfo);
+    }
+  }
+  
+  private String buildNamePath(String volumeName, String bucketName) {
+    final StringBuilder builder = new StringBuilder();
+    builder.append(OM_KEY_PREFIX)
+        .append(volumeName)
+        .append(OM_KEY_PREFIX)
+        .append(bucketName)
+        .append(OM_KEY_PREFIX);
+    return builder.toString();
+  }
+
+  private String buildIdPath(long volumeId, long bucketId) {
+    final StringBuilder builder = new StringBuilder();
+    builder.append(OM_KEY_PREFIX)
+        .append(volumeId)
+        .append(OM_KEY_PREFIX)
+        .append(bucketId)
+        .append(OM_KEY_PREFIX);
+    return builder.toString();
+  }
+  
+  private void repairCount() throws Exception {
+    LOG.info("Starting quota repair for all keys, files and directories");
+    try {
+      nameBucketInfoMap.keySet().stream().forEach(e -> keyCountMap.put(e,
```

Comment:
this is common function for key, file and directory, it should be context-less. so to move this inside, need pass nameBucketInfoMap/idBucketInfoMap to method. Since the number of bucket is few thousands and memory operation takin less than second, this movement will not provide benefits. So its avoided.